packages/nango-yaml/lib/modelsParser.ts

```diff
@@ -0,0 +1,200 @@
+import type { NangoModel, NangoModelField, NangoYamlModel, NangoYamlModelFields } from '@nangohq/types';
+import { getNativeDataType, getPotentialTypeAlias, isDisallowedType, shouldQuote } from './helpers.js';
+import { ParserError, ParserErrorCycle, ParserErrorDataSyntax, ParserErrorExtendsNotFound, ParserErrorInvalidModelName } from './errors.js';
+
+export class ModelsParser {
+    parsed = new Map<string, NangoModel>();
+    references = new Set<string>();
+    errors: ParserError[] = [];
+    warnings: ParserError[] = [];
+
+    raw: NangoYamlModel;
+
+    constructor({ raw }: { raw: NangoYamlModel }) {
+        this.raw = raw;
+    }
+
+    parseAll() {
+        for (const [name, fields] of Object.entries(this.raw)) {
+            if (shouldQuote(name)) {
+                this.errors.push(new ParserErrorInvalidModelName({ model: name, path: [name] }));
+                continue;
+            }
+            if (this.parsed.has(name)) {
+                continue;
+            }
+
+            this.parseOne({ name, fields });
+        }
+    }
+
+    get(name: string): NangoModel | undefined {
+        const parsed = this.parsed.get(name);
+        if (!parsed) {
+            return undefined;
+        }
+
+        return parsed;
+    }
+
+    parseOne({ name, fields }: { name: string; fields: NangoYamlModelFields }): void {
+        const parsed = this.parseFields({ fields, parent: name });
+        if (parsed) {
+            this.parsed.set(name, { name, fields: parsed });
+        }
+    }
+
+    ifModelParse({ name, parent }: { name: string; parent: string }): true | false {
+        if (this.references.has(`${parent}-${name}`) || parent === name) {
+            this.references.add(`${parent}-${name}`);
+            this.warnings.push(new ParserErrorCycle({ name }));
+            return true;
+        }
+
+        // Previously parsed skip things
+        const has = this.parsed.get(name);
+        if (has) {
+            return true;
+        }
+
+        // Model does not exists but that could just mean string literal
+        if (!this.raw[name]) {
+            this.warnings.push(
+                new ParserError({ code: 'model_not_found_fallback', message: `Model "${name}" is not defined, using as string literal`, path: [parent, name] })
+            );
+            return false;
+        }
+
+        // At this point we are sure it's a Model
+        this.parseOne({ name, fields: this.raw[name]! });
+        return true;
+    }
+
+    parseFields({ fields, parent }: { fields: NangoYamlModelFields; parent: string }): NangoModelField[] {
```

Comment:
This is the main function to parse models and handle all the edge cases regarding types